lib/format_parser.rb

```diff
@@ -30,24 +39,30 @@ def self.parse_http(url)
     parse(cached_io)
   end
 
-  def self.parse(io)
+  def self.parse(io, **opts, &proc)
+    config = parse_config(**opts, &proc)
     # If the cache is preconfigured do not apply an extra layer. It is going
     # to be preconfigured when using parse_http.
     io = Care::IOWrapper.new(io) unless io.is_a?(Care::IOWrapper)
+    results = []
 
     # Always instantiate parsers fresh for each input, since they might
     # contain instance variables which otherwise would have to be reset
     # between invocations, and would complicate threading situations
-    parsers = @parsers.map(&:new)
+    parsers = @parsers.select do |p|
+      !(p.natures & config.natures).empty? && !(p.formats & config.formats).empty?
+    end.map(&:new)
 
     parsers.each do |parser|
       # We need to rewind for each parser, anew
       io.seek(0)
       # Limit how many operations the parser can perform
       limited_io = ReadLimiter.new(io, max_bytes: 512*1024, max_reads: 64*1024, max_seeks: 64*1024)
       begin
-        if info = parser.information_from_io(limited_io)
-          return info
+        if info = parser.call(limited_io)
+          results << info
+          # Return early if the limit was hit.
+          return results if config.limit == results.length
```

Comment:
Can there be a case when we want something else than "only the first one" and "all of the ones matched"?

Reply:
Both cases are now supported. 

```ruby
FormatParser.parse(f, limit: 1)
FormatParser.parse(f) # it trust that a parser only returns one value
```